clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -808,7 +808,8 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
                 }
                 log.debug("Subscribed to topic(s): {}", Utils.join(topics, ", "));
                 this.subscriptions.subscribe(topics, listener);
-                metadata.setTopics(subscriptions.groupSubscription());
+                this.metadata.needMetadataForAllTopics(true);
```

Comment:
Hmm.. But when you call `SubscriptionState.subscribe()`, we should set the flag saying that we need a rebalance. So the next call to poll() will force us to rebalance and we won't be able to return any data for an unsubscribed partition. Is that not right?